### ISSUE TITLE
Graphics very slow on AMD G-Series Jaguar/Radeon HD 8400E GPU

### ISSUE THREAD

**Issue by nkcyham** [voidlinux/void-packages#2916]
Hi

I eventually replaced my failed desktop box with a fit-PC4 from Compulab, I deleted the original Linux Mint installation and installed Void. All is well, except that graphics are very slow, at least with one of
my applications, xnec2c, which calculates antenna radiation patterns and displays in a window as a wire frame. The slowness is apparent when the wireframe display is rotated by dragging it with the pointer. Normally the wire frame follows the pointer in real time but with Void on the fitPC4 the wire frame movement can only follow the pointer if it is moved very slowly.

I downloaded a Debian Live Amateur Radio spin which contains xnec2c and tried it there, the wire frame movement is quite fast, following the pointer. I compared the time it takes for xnec2c to complete the antenna simulation and it is about 20 secs on both Void and Debian, so it is not an issue with the huge amount of floating point calculations needed. It seems to be an issue with the GPU system, perhaps Xorg or driver issue, possibly compile time configuration.

Further data: I installed Debian Live Amateur Radio on a spare partition and the above situation changed: In order to enable iwlwifi, I copied non-free firmware from void to debian (from-to /lib/firmware): The result was that xnec2c slowed down considerably, just as in Void. Eventually I found that by deleting the KABINI/kabini files in lib/firmware/radeon and rebooting debian, the problem is fixed. So its the radeon firmware that causes the problem!

I tried to do the same thing in Void, but rebuilding the initrd with dracut I then cannot boot, as I have apparently no video driver for the console? Both amdgpu and radeom drivers will not work without the kabini firmware, although debian seems to work without them, apparently a lot better! If I can find a way to boot void without radeon/kabini firmware, I could verify the problem on void too - but I couldn't!

Any ideas please?

p.s. same problem with archlinux.

**Comment by pullmoll** [comment]
All seems well with yout Xorg.0.log, GL is detected: `GLX: Initialized DRI2 GL provider for screen 0`

I see no official package `xnec2c`, which is why I assume you created your own package and template? In that case make sure you install the required development libaries such as `MesaLib-devel` or perhaps `glu-devel` and watch out, if in the `configure` stage the GL or GLU drivers are detected.

To see if `xnec2c` links against e.g. `/usr/lib/libGL.so.1` you can run `ldd /usr/bin/xnec2c` - or whereever you installed the binary.

**Comment by pullmoll** [comment]
I'm trying to package xnec2c. It does not link against libGL, so perhaps it tries to dynamically load the GL functions. I'll see if I can get a complete package whipped up which also installs the dependencies, if any.

**Comment by nkcyham** [comment]
On Sat, 11/7/15, Jürgen Buchmüller <notifications@github.com> wrote:

 >I'm trying to package xnec2c.
 >It does not link against libGL, so perhaps it tries to
 >dynamically load the GL functions. I'll see if I can get
 >a complete package whipped up which also installs the
 >dependencies, if any.
Rather embarrassing, I wrote xnec2c (started a long time ago)
and are forgetting the details. It doesn't use libGL, not directly.
All graphics are produced using GTK+-2.0 functions and Cairo
functions. xnec2c is installed here directly from my own source
code so all dependencies are available.

Thanks for working on packaging xnec2c, its included in a few
distros so Void should have it too! ;-)

--
Best Regards
Neoklis - Ham Radio Call:5B4AZ
http://www.5b4az.org/

**Comment by nkcyham** [comment]
On Sat, 11/7/15, Jürgen Buchmüller <notifications@github.com> wrote:

 All seems well with yout Xorg.0.log, GL is detected: GLX:
 Initialized DRI2 GL provider for screen 0
 
> I see no official package xnec2c, which is why I assume you
> created your own package and template?
No, I installed from source code, since I wrote xnec2c myself.

> In that case make sure you install the required development
> libaries such as MesaLib-devel or perhaps...
Well, I install xnec2c from the source and all dependencies are
available. All graphics are done mostly by Cairo so I have no 
direct involvement with MesaLib or LibGL etc.

Anyhow, thanks for trying! :-)
 
 —
 Reply to this email directly or view
 it on GitHub.
 
 
   
   

--
Best Regards
Neoklis - Ham Radio Call:5B4AZ
http://www.5b4az.org/

**Comment by pullmoll** [comment]
For me rotating the wireframes goes rather fast. I see no slowdown. I have no idea where to look now. Someone else perhaps?

**Comment by nkcyham** [comment]
On Sat, 11/7/15, Jürgen Buchmüller <notifications@github.com> wrote:

> For me rotating the wireframes goes rather fast. I see no slowdown. 
> I have no idea where to look now. Someone else perhaps?
 OK Jürgen, it must be related to the hardware. Since it works very well
in Debian, its likely some compile time config of either the kernel or the
GTK+ or Cairo that does it. I will try here to see if I can isolate something.

Thanks for trying.

--
Best Regards
Neoklis - Ham Radio Call:5B4AZ
http://www.5b4az.org/

**Comment by nkcyham** [comment]
OK

I tried some ideas and the problem seems to be the Cairo lib or its dependencies.
The wire frames are drawn using the following code: (draw.c, 297)

/* Cairo_Draw_Line()
 *
 * Draws a line between to x,y co-ordinates
 */
  void
Cairo_Draw_Line( cairo_t *cr, int x1, int y1, int x2, int y2 )
{
  cairo_move_to( cr, (double)x1, (double)y1 );
  cairo_line_to( cr, (double)x2, (double)y2 );
  cairo_stroke( cr );
} /* Cairo_Draw_Line() */

If I comment out cairo_stroke( cr ); then there is no wireframe displayed,
only the xyz axis and response to dragging the pointer is much faster. I
also tried moving the cairo_stroke( cr ); line to the end of the function that
draws the wire frame (draw_radiation.c, 86), the wireframe is drawn much
faster but only the last color used in cairo's context is shown.

I suppose this means that somewhere in the chain of libraries from cairo to
Xorg my GPU hardware is not handled properly, most likely hardware
acceleration is not used properly.

Anyhow, its not very important or urgent but I am curious as to what is wrong.

--
Best Regards
Neoklis - Ham Radio Call:5B4AZ
http://www.5b4az.org/

**Comment by nkcyham** [comment]
Further data: I installed Debian Live Amateur Radio on a spare partition and the above situation changed: In order to enable iwlwifi, I copied non-free firmware from void to debian (from-to /lib/firmware): The result was that xnec2c slowed down considerably, just as in Void. Eventually I found that by deleting the KABINI/kabini files in lib/firmware/radeon and rebooting debian, the problem is fixed. So its the radeon firmware that causes the problem!

I tried to do the same thing in Void, but rebuilding the initrd with dracut I then cannot boot, as I have apparently no video driver for the console? Both amdgpu and radeom drivers will not work without the kabini firmware, although debian seems to work without them, apparently a lot better! If I can find a way to boot void without radeon/kabini firmware, I could verify the problem on void too - but I couldn't!

Any ideas please?

p.s. same problem with archlinux.

**Comment by nkcyham** [comment]
It now looks like mesa+libGL does not or cannot use hardware acceleration when the KABINI firmware blobs are loaded, slowing down a lot Cairo rendering. I veryfied this by:

1. Removing the kabini blobs from /lib/firmware/ and rebuilding the initrd. The result is that
Cairo rendering is much faster and processor workload is very much reduced. However the console goes blank but I can log in blind and start my X desktop.

2. With the kabini blobs in place, I installed nvidia-libs in place of libGL and Cairo rendering again is much faster and processor workload much reduced.

3. I installed catalyst-libs in place of libGL and again as above I have much faster rendering etc.

So I will file a bug report to the mesa3d devs, hopefully they may do something about it.
